state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,274 @@
+							.checked_sub(&electoral_access.unsynchronised_state()?)
+							.expect("current block number is always greater than when apicall was last created")
+							.into() >= MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS)
+				{
+					let accounts_to_close: Vec<_> = known_accounts
+					.witnessed_open_accounts
+					.drain(..sp_std::cmp::min(
+						known_accounts.witnessed_open_accounts.len(),
+						MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES
+					))
+					.collect();
+					match Hook::close_accounts(accounts_to_close.clone()) {
+						Ok(()) => {
+							known_accounts.closure_initiated_accounts.extend(accounts_to_close);
+							electoral_access.set_unsynchronised_state(*current_block_number)?;
+						},
+						Err(e) => {
+							log::error!(
+								"failed to close accounts: {:?}",
+								e
+							);
+							known_accounts.witnessed_open_accounts.extend(accounts_to_close);
+						},
+					}
+				}
+				electoral_access.new_election((), known_accounts, ())?;
+			}
+		} else {
+			electoral_access.new_election(
+				(),
+				SolanaVaultSwapsKnownAccounts {
+					witnessed_open_accounts: Vec::new(),
+					closure_initiated_accounts: BTreeSet::new(),
+				},
+				(),
+			)?;
+		}
+
+		Ok(())
+	}
+
+	fn check_consensus<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier: ElectionIdentifier<Self::ElectionIdentifierExtra>,
+		_election_access: &ElectionAccess,
+		_previous_consensus: Option<&Self::Consensus>,
+		consensus_votes: ConsensusVotes<Self>,
+	) -> Result<Option<Self::Consensus>, CorruptStorageError> {
+		let num_authorities = consensus_votes.num_authorities();
+		let success_threshold = success_threshold_from_share_count(num_authorities);
+		let active_votes = consensus_votes.active_votes();
+		let num_active_votes = active_votes.len() as u32;
+		Ok(if num_active_votes >= success_threshold {
+			let mut counts_votes = BTreeMap::new();
+			let mut counts_new_accounts = BTreeMap::new();
+			let mut counts_confirm_closed_accounts = BTreeMap::new();
+
+			for vote in active_votes {
+				counts_votes.entry(vote).and_modify(|count| *count += 1).or_insert(1);
+			}
+
+			counts_votes.iter().for_each(|(vote, count)| {
+				vote.new_accounts.iter().for_each(|new_account| {
+					counts_new_accounts
+						.entry(new_account)
+						.and_modify(|c| *c += *count)
+						.or_insert(*count);
+				});
+				vote.confirm_closed_accounts.iter().for_each(|confirm_closed_account| {
+					counts_confirm_closed_accounts
+						.entry(confirm_closed_account)
+						.and_modify(|c| *c += *count)
+						.or_insert(*count);
+				});
+			});
+
+			counts_new_accounts.retain(|_, count| *count >= success_threshold);
+			let new_accounts = counts_new_accounts.into_keys().cloned().collect::<BTreeSet<_>>();
+			counts_confirm_closed_accounts.retain(|_, count| *count >= success_threshold);
+			let confirm_closed_accounts =
+				counts_confirm_closed_accounts.into_keys().cloned().collect::<BTreeSet<_>>();
```

Comment:
for each of new_account and confirm_closed accounts there's dupe logic - could factor out